### DISCUSSION TITLE
Would Nextjs Api solve my CORS issue?

### DISCUSSION THREAD

**Discussion by nonoumasy** [vercel/next.js#40291] [Help]
I'm working on a project and I'm using this library dom-to-image. I'm getting CORS issue with this and one of the ways to fix it is to setup a proy server. I can create one using Node but was wondering if I use Nextjs already has a solution for this and I'll just use Nextjs. I'm currently just using vanilla Reactjs. 
If I use Nextjs api to fetch the data from Firestore and then on my frontend, I would fetch it from my Next API route, wouldn't this essentially fix the CORS issue I'm having?
Thanks in advance.

**Comment by icyJoseph** [comment]
Hi,

Well, it might, I'd try it out firsts, with a pet project, it should go quite fast to setup and test it. 

CORS can be triggered by a number of reasons, but essentially what you describe should do the trick.

Because the `api/routes` are sitting on the same domain, you can make a same-domain request to them, and have that one in turn communicate with whatever other services you want to use. 

What's even nicer is that you could code the `api/route` to just proxy the request through to the intended target, checkout https://maxschmitt.me/posts/next-js-api-proxy/, though I've always found myself doing tweaks to these 2 steps solutions.

There are two caveats though, timeouts and payload sizes. API routes are lightweight and are limited on both the time they can run (depends on your host but Vercel allows about 10 seconds, IIRC), and are also limited on the payload size they can allow, at about 4.5MB, so any sufficiently large image might cause trouble anyway. https://vercel.com/docs/concepts/limits/overview

Ideally you should be able to tell Firestore to allow your domain for the type of requests you are doing.

Finally, what do you mean with vanilla React.js? Do you mean CRA/Vite?

**Reply by nonoumasy** [reply]
Thanks @icyJoseph 

Yes I meant CRA when I said vanilla React.js (just to differentiate between Nextjs)

so I refactored using Nextjs.
I'm fetching the data from the backend using firebase-admin
and I'm fetching this data from the frontend.
Everything works as before, but I still got the CORS error.

So then, I took a look at:
https://maxschmitt.me/posts/next-js-api-proxy/
and installed the library and then added the code below to the dynamic API route.
I must have missed a step because now it doesn't fetch my data and the CORS problem is still there. 
Maybe, I put the wrong value for the target property?

```
  proxy.web(req, res, {
    target: "/api/*",
    changeOrigin: true,
  })
```
I also tried:

```
  proxy.web(req, res, {
    target: "http://localhost:3000/api/story/[storyId]",
    changeOrigin: true,
  })
```

Whole code:
```
import admin from "firebase-admin"
import httpProxy from "http-proxy"
import serviceAccount from "../../../ServiceAccountKey.json"

if (!admin.apps.length) {
  admin.initializeApp({
    credential: admin.credential.cert(serviceAccount),
    name: "timeline2",
  })
}

const db = admin.firestore()
const proxy = httpProxy.createProxyServer()

// Make sure that we don't parse JSON bodies on this route:
export const config = {
  api: {
    bodyParser: false,
  },
}

export default function handler(req, res) {

  const { storyId } = req.query

//proxy settings
  proxy.web(req, res, {
    target: "/api/*",
    changeOrigin: true,
  })

//fetches dynamic data based on "storyId"
  const getEvent = async () => {
    const data = await db
      .collection("stories")
      .doc(storyId)
      .collection("events")
      .get()
    res.json(
      data.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }))
    )
  }
  getEvent()
}
```
